File: backend/neural_mri/schemas/causal_trace.py
from __future__ import annotations

from pydantic import BaseModel


class CausalTraceRequest(BaseModel):
    clean_prompt: str
    corrupt_prompt: str
    target_token_idx: int = -1  # default: last token


class CausalTraceCell(BaseModel):
    component: str  # "embed", "blocks.0.attn", "blocks.0.mlp", ...
    layer_idx: int  # -1 for embed
    component_type: str  # "attn" | "mlp" | "embed"
    recovery_score: float  # 0-1


class CausalTraceResult(BaseModel):
    model_id: str
    clean_prompt: str
    corrupt_prompt: str
    target_token_idx: int
    clean_prediction: str
    corrupt_prediction: str
    cells: list[CausalTraceCell]
    n_layers: int
    metadata: dict

File: frontend/src/types/causalTrace.ts
export interface CausalTraceCell {
  component: string;
  layer_idx: number;
  component_type: 'attn' | 'mlp' | 'embed';
  recovery_score: number;
}

export interface CausalTraceResult {
  model_id: string;
  clean_prompt: string;
  corrupt_prompt: string;
  target_token_idx: number;
  clean_prediction: string;
  corrupt_prediction: string;
  cells: CausalTraceCell[];
  n_layers: number;
  metadata: Record<string, unknown>;
}
